src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcServerTransport.cs

```diff
@@ -0,0 +1,163 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.IO.Pipes;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    internal abstract class IpcServerTransport : IDisposable
+    {
+        public static IpcServerTransport Create(string transportPath)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return new WindowsPipeServerTransport(transportPath);
+            }
+            else
+            {
+                return new UnixDomainSocketServerTransport(transportPath);
+            }
+        }
+
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+        }
+
+        public abstract Task<Stream> AcceptAsync(CancellationToken token);
+    }
+
+    internal sealed class WindowsPipeServerTransport : IpcServerTransport
+    {
+        private const string PipePrefix = @"\\.\pipe\";
+
+        private bool _disposed = false;
+        private NamedPipeServerStream _stream;
+
+        private readonly string _pipeName;
+
+        public WindowsPipeServerTransport(string pipeName)
+        {
+            _pipeName = pipeName.StartsWith(PipePrefix) ? pipeName.Substring(PipePrefix.Length) : pipeName;
+            CreateNewPipeServer();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                if (disposing)
+                {
+                    _stream.Dispose();
+                }
+                _disposed = true;
+            }
+        }
+
+        public override async Task<Stream> AcceptAsync(CancellationToken token)
+        {
+            NamedPipeServerStream connectedStream;
+            try
+            {
+                await _stream.WaitForConnectionAsync(token);
+
+                connectedStream = _stream;
+            }
+            finally
+            {
+                CreateNewPipeServer();
+            }
+            return connectedStream;
+        }
+
+        private void CreateNewPipeServer()
+        {
+            _stream = new NamedPipeServerStream(_pipeName, PipeDirection.InOut, 10, PipeTransmissionMode.Byte, PipeOptions.Asynchronous);
+        }
+    }
+
+    internal sealed class UnixDomainSocketServerTransport : IpcServerTransport
+    {
+        private bool _disposed = false;
+
+        private readonly string _path;
+        private readonly Socket _socket;
+
+        public UnixDomainSocketServerTransport(string path)
+        {
+            _path = path;
+            _socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
+            _socket.Bind(PidIpcEndpoint.CreateUnixDomainSocketEndPoint(path));
+            _socket.Listen(255);
```

Comment:
nit: I would put backlog and instance limits as constants so that they are easily visible/changable

Reply:
I'm thinking that I could combine the socket backlog and the named pipe instance count into the same concept for the reverse server because I think they represent similar concepts as implemented here. Then these can be made a parameter to the reversed server rather than being something fixed.